### ISSUE TITLE
Examples don't work on macOS CI runner

### ISSUE THREAD

**Issue by AlexeiDrake** [tweag/rules_nixpkgs#176]
**Describe the bug**
The examples CI job does not work on any version of the macOS runner.

**To Reproduce**
Add `macos-11` or `macos-10.15` to the list of runners for the examples job.

**Expected behavior**
CI jobs should pass.

**Additional context**
1. Our examples pin nixpkgs, ideally to a release to avoid cache misses. At the time of writing, the most obvious options are 21.05 and 21.11.
2. The Go toolchain requires at least Bazel 4.2.x., which is present in 21.11, but not in 21.05 (4.0.0).
3. We must therefore pin against at least 21.11.
4. nixpkgs at 21.11 seems to have broken the Darwin stdenv for macos-10.15 with this commit: https://github.com/NixOS/nixpkgs/commit/a634fbe065c1800ebb1e1e1dd7a691115280493d
5. We must therefore use at least macos-11.
6. macos-11 CI support is broken and tracked with this bug: [tweag/rules_nixpkgs#175](https://github.com/tweag/rules_nixpkgs/issues/175)

**Comment by aherrmann** [comment]
Thanks for writing up the context!

> [...] ideally to a release to avoid cache misses [...]
> 3\. We must therefore pin against at least 21.11.

This is not strictly true. Nixpkgs is continuously built on hydra and pushed to the nixpkgs community cache in between releases as well. To find a cached nixpkgs revision that provides that bazel version you can
* browse to the [nixpkgs hydra job](https://hydra.nixos.org/jobset/nixpkgs/trunk)
* [search](https://hydra.nixos.org/jobset/nixpkgs/trunk#tabs-jobs) for [bazel_4 jobs](https://hydra.nixos.org/job/nixpkgs/trunk/bazel_4.x86_64-darwin/all)
* find one that provides Bazel version 4.2.x but predates https://github.com/NixOS/nixpkgs/commit/a634fbe065c1800ebb1e1e1dd7a691115280493d.
* and lookup the nixpkgs revision in its [inputs](https://hydra.nixos.org/build/153268139#tabs-buildinputs)

E.g. https://hydra.nixos.org/build/153268139 seems to fit the bill: It provides Bazel 4.2.1 and [not contain the offending line in stdenv](https://github.com/NixOS/nixpkgs/blob/a3c4956cf9cb921d61b4a5c30df6ef1c07d2fae4/pkgs/stdenv/generic/setup.sh#L1-L3). A quick local test shows bazel being fetched from cache
```
$ nix-build https://github.com/NixOS/nixpkgs/archive/a3c4956cf9cb921d61b4a5c30df6ef1c07d2fae4.tar.gz -A bazel_4
```

See also [here](https://discourse.nixos.org/t/steps-to-find-a-commit-with-cached-version-of-package/5674) for some hints on how to find cached revisions.

**Comment by AlexeiDrake** [comment]
Indeed that wording is a bit too strict. I think pinning to releases is a lot more representative of real-world usage however; if someone uses the example in the wild and wishes to update nixpkgs to get an updated toolchain, it might be a nasty surprise when it fails on macOS.

**Comment by YorikSar** [comment]
Rust example will be fixed in #187.
